**AYGAZ GÖRÜNTÜ İŞLEME BOOTCAMP PROJESİ**

Hazırlayan : Duygu Bulut


# 1. Importing and Downloading Dataset:
**Kaggle dataset download:**

This code imports the kagglehub library and uses it to download the "Animals with Attributes 2" dataset from Kaggle. After successful download, a message is printed to confirm that the dataset is ready for use.

In [8]:
# IMPORTANT: RUN THIS CELL IN ORDER TO IMPORT YOUR KAGGLE DATA SOURCES,
# THEN FEEL FREE TO DELETE THIS CELL.
# NOTE: THIS NOTEBOOK ENVIRONMENT DIFFERS FROM KAGGLE'S PYTHON
# ENVIRONMENT SO THERE MAY BE MISSING LIBRARIES USED BY YOUR
# NOTEBOOK.
import kagglehub
rrebirrth_animals_with_attributes_2_path = kagglehub.dataset_download('rrebirrth/animals-with-attributes-2')

print('Data source import complete.')


100%|██████████| 13.0G/13.0G [02:00<00:00, 115MB/s]

Extracting files...


Data source import complete.


# 2. Preparing the Dataset:



**Organizing Dataset Paths:**

This section organizes the dataset by traversing the directory structure to find images of specific animal classes.

It creates a dictionary (image_paths) to store file paths for each animal.
It iterates through the directory and checks if a folder corresponds to an animal class in the specified list.
For matched folders, it collects image file paths and adds them to the dictionary.
Finally, it prints the count of images for each class to verify the data.

In [9]:
import os

# Initialize the dictionary to hold paths for each animal
image_paths = {}

# Base directory for your dataset
base_path = "/root/.cache/kagglehub/datasets/rrebirrth/animals-with-attributes-2/versions/1/Animals_with_Attributes2/JPEGImages/"

# List of animals to search for
animals = ["collie", "dolphin", "elephant", "fox", "moose", "rabbit", "sheep", "squirrel", "giant+panda", "polar+bear"]

# Traverse the directory structure
for dirname, _, filenames in os.walk(base_path):
    for animal in animals:
        # Check if the current directory contains the animal's name
        if animal in dirname:
            # Initialize the list if the animal is encountered for the first time
            if animal not in image_paths:
                image_paths[animal] = []
            # Add all image paths for the current animal
            for filename in filenames:
                image_paths[animal].append(os.path.join(dirname, filename))

# Print the paths for verification
for animal, paths in image_paths.items():
    print(f"{animal}: {len(paths)} images")

giant+panda: 874 images
polar+bear: 868 images
rabbit: 1088 images
elephant: 1038 images
collie: 1028 images
squirrel: 1200 images
sheep: 1420 images
fox: 664 images
moose: 704 images
dolphin: 946 images


**Transferring Dataset to Google Drive**

This section connects to Google Drive and copies the dataset for easier access and organization.

The Drive is mounted to the Colab environment.
A target folder is created in Google Drive to store images by class.
The code iterates through the dataset directory, copying images from the source directory to the target folder, organizing them by class.
A success message is printed to confirm the completion of the transfer.

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [12]:
import os
import shutil

# source dataset path
source_base_path = "/root/.cache/kagglehub/datasets/rrebirrth/animals-with-attributes-2/versions/1/Animals_with_Attributes2/JPEGImages/"

# selected animal classes
animals = ["collie", "dolphin", "elephant", "fox", "moose", "rabbit", "sheep", "squirrel", "giant+panda", "polar+bear"]

# create a target folder in drive to use the dataset more easily
drive_dataset_path = "/content/drive/My Drive/bootcamp_project/animal_dataset/"
os.makedirs(drive_dataset_path, exist_ok=True)

# copy data containing animal classes
for dirname, _, filenames in os.walk(source_base_path):
    for animal in animals:
        if animal in dirname:
           # the target folder will be organized by class name
            target_dir = os.path.join(drive_dataset_path, animal)
            os.makedirs(target_dir, exist_ok=True)
            for filename in filenames:
                source_file = os.path.join(dirname, filename)
                target_file = os.path.join(target_dir, filename)
                # copy the file
                shutil.copy2(source_file, target_file)

print(f"Selected classes successfully copied to: {drive_dataset_path}")

Selected classes successfully copied to: /content/drive/My Drive/bootcamp_project/animal_dataset/


**Listing Folders and Image Count**

This code lists all animal classes and prints the number of images for each class in the dataset. It ensures that only directories (not files) are processed and counts the number of files in each folder.

In [13]:
import os

# base folder path where the dataset is located
dataset_path = "/content/drive/My Drive/bootcamp_project/animal_dataset/"

# list folders and image numbers for printing
for animal_class in sorted(os.listdir(dataset_path)):
    class_path = os.path.join(dataset_path, animal_class)
    if os.path.isdir(class_path):  # check the folders
        image_count = len([f for f in os.listdir(class_path) if os.path.isfile(os.path.join(class_path, f))])
        print(f"{animal_class}: {image_count} images")

collie: 1028 images
dolphin: 946 images
elephant: 1038 images
fox: 664 images
giant+panda: 874 images
moose: 704 images
polar+bear: 868 images
rabbit: 1088 images
sheep: 1420 images
squirrel: 1200 images


# 2.1. Create balanced dataset

This script balances the dataset by ensuring each animal class has an equal number of images (650 in this case).

For each class, it selects the first 650 images.
It creates a new directory structure in a specified path to store the balanced dataset.
Images are copied from the original dataset to the new balanced dataset directory.

In [14]:
import os
import shutil

# This piece of code is to initially take the first 650 images of each desired animal class and save them to a new folder on the drive

# current dataset path
dataset_path = "/content/drive/My Drive/bootcamp_project/animal_dataset/"

# create new balanced dataset path
balanced_dataset_path = "/content/drive/My Drive/bootcamp_project/balanced_animal_dataset/"
os.makedirs(balanced_dataset_path, exist_ok=True)

# target number per image
target_count = 650

# moving selected files from each class to a new folder
for animal_class in sorted(os.listdir(dataset_path)):
    class_path = os.path.join(dataset_path, animal_class)
    if os.path.isdir(class_path):  # process folders only
        # get all image files in folder in order
        all_images = sorted([f for f in os.listdir(class_path) if os.path.isfile(os.path.join(class_path, f))])
        # select first 650 images
        selected_images = all_images[:target_count]
        # create a folder for the class in the new destination folder
        target_class_path = os.path.join(balanced_dataset_path, animal_class)
        os.makedirs(target_class_path, exist_ok=True)
        # copy selected files to new folder
        for image in selected_images:
            source_file = os.path.join(class_path, image)
            target_file = os.path.join(target_class_path, image)
            shutil.copy2(source_file, target_file)

        print(f"{animal_class}: {len(selected_images)} images copied")

print("Balanced dataset creation complete.")

collie: 650 images copied
dolphin: 650 images copied
elephant: 650 images copied
fox: 650 images copied
giant+panda: 650 images copied
moose: 650 images copied
polar+bear: 650 images copied
rabbit: 650 images copied
sheep: 650 images copied
squirrel: 650 images copied
Balanced dataset creation complete.


# 2.2. Resizing and Normalizing Dataset

This function resizes and normalizes images in the dataset.

It resizes images to a fixed size (e.g., 64x64 pixels).
It normalizes pixel values to a range of [0, 1].
Resized images are saved in a new directory, maintaining the original directory structure.

In [15]:
'''
To normalize all images to the same size.
Resize images to the appropriate size according to the input layer of the model.
'''
import os
import cv2
import numpy as np

# image resizing and saving function
def resize_and_save_images(input_path, output_path, image_size=(128, 128)):
    if not os.path.exists(output_path):
        os.makedirs(output_path)
    # loop through all image files
    for root, dirs, files in os.walk(input_path):
        for file in files:
            if file.endswith(('.jpg', '.jpeg', '.png')):
                file_path = os.path.join(root, file)
                input_image = cv2.imread(file_path)
                # resize if input image is not none
                if input_image is not None:
                    img_resized = cv2.resize(input_image, image_size)
                    img_normalized = img_resized / 255.0  # normalization
                    relative_path = os.path.relpath(root, input_path)
                    save_dir = os.path.join(output_path, relative_path)
                    # create save_dir if it doesn't exist
                    if not os.path.exists(save_dir):
                        os.makedirs(save_dir)
                    # save images to output_file_path
                    output_file_path = os.path.join(save_dir, file)
                    cv2.imwrite(output_file_path, (img_normalized * 255).astype(np.uint8))

# input and output dataset paths
input_dataset_path = "/content/drive/My Drive/bootcamp_project/balanced_animal_dataset/"
output_dataset_path = "/content/processed_data"

# call the function
resize_and_save_images(input_dataset_path, output_dataset_path, image_size=(64, 64))

# 2.3. Preparation and splitting of data for training and test dataset

This code prepares the dataset for training:

Loads images and their labels, associating each class name with a unique numerical label.
Normalizes images.
Splits the dataset into training (70%) and test (30%) subsets using train_test_split.

In [16]:
import os
import numpy as np
from sklearn.model_selection import train_test_split

# preparation of data and labels
def load_processed_dataset(data_path):
    images = []
    labels = []
    class_names = sorted(os.listdir(data_path))  # classes in alphabetical order
    class_to_label = {cls_name: idx for idx, cls_name in enumerate(class_names)}

    for cls_name in class_names:
        class_path = os.path.join(data_path, cls_name)
        if os.path.isdir(class_path):
            for file in os.listdir(class_path):
                if file.endswith(('.jpg', '.jpeg', '.png')):
                    file_path = os.path.join(class_path, file)
                    img = cv2.imread(file_path)  # Image resized and normalized
                    if img is not None:
                        images.append(img / 255.0)  # make sure it is normalized for validation
                        labels.append(class_to_label[cls_name])

    return np.array(images), np.array(labels), class_names

# load the dataset
processed_dataset_path = "/content/processed_data"  # Processed data path
X, y, class_names = load_processed_dataset(processed_dataset_path)

# splitting into training and test data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# printing results
print(f"Toplam veri sayısı: {len(X)}")
print(f"Eğitim veri sayısı: {len(X_train)}")
print(f"Test veri sayısı: {len(X_test)}")
print(f"Sınıflar: {class_names}")

Toplam veri sayısı: 6500
Eğitim veri sayısı: 4550
Test veri sayısı: 1950
Sınıflar: ['collie', 'dolphin', 'elephant', 'fox', 'giant+panda', 'moose', 'polar+bear', 'rabbit', 'sheep', 'squirrel']


# 2.4. Data manipulation and augmentation

This code performs data augmentation to artificially expand the training dataset:

Applies random transformations (rotation, zoom, flip, and translation) to the training images.
Creates a specified number of augmented versions of each image.
Combines the original and augmented datasets, increasing the variety in the training data to improve model robustness.

In [17]:
import tensorflow as tf
# Data augmentation settings for tensorflow layers
data_augmentation = tf.keras.Sequential([
    tf.keras.layers.RandomRotation(0.1),  # ±10% rotate
    tf.keras.layers.RandomZoom(0.1),      # ±10% zoom
    tf.keras.layers.RandomFlip("horizontal"),  # horizontal flip
    tf.keras.layers.RandomTranslation(0.1, 0.1),  # width/height shift
])

# applying data augmentation to the training set
def augment_images_tf(X_train, y_train, num_augmented=5):
    augmented_images = []
    augmented_labels = []

    for img, label in zip(X_train, y_train):
        img = tf.expand_dims(img, axis=0)  # # to make the format suitable for TensorFlow
        for _ in range(num_augmented):
            augmented_img = data_augmentation(img, training=True)
            augmented_images.append(augmented_img[0].numpy())
            augmented_labels.append(label)

    return np.array(augmented_images), np.array(augmented_labels)

# data augmentation process
# num_augmented : Specified number of increments for each image
augmented_X_train, augmented_y_train = augment_images_tf(X_train, y_train, num_augmented=5)

# combining the results
X_train_augmented = np.concatenate((X_train, augmented_X_train), axis=0)
y_train_augmented = np.concatenate((y_train, augmented_y_train), axis=0)
# print shape of the new dataset
print(f"Orijinal eğitim veri boyutu: {X_train.shape}")
print(f"Artırılmış eğitim veri boyutu: {X_train_augmented.shape}")

Orijinal eğitim veri boyutu: (4550, 64, 64, 3)
Artırılmış eğitim veri boyutu: (27300, 64, 64, 3)


# 3. Designing the CNN Model:

# 3.1 Building a Convolutional Neural Network (CNN) with TensorFlow

This CNN is designed for multi-class classification of images (10 classes). It processes the input through convolutional, pooling, and fully connected layers to extract features, reduce dimensionality, and classify the images based on their features.

This code constructs a CNN model for classifying images into 10 classes. Here's a breakdown of its components:
1. Input Layer and First Convolutional Block

Conv2D(32, (3, 3), activation='relu'):
Adds a convolutional layer with 32 filters, each of size 3x3.
Applies the ReLU activation function to introduce non-linearity.
input_shape=(64, 64, 3):
Specifies the input image dimensions (64x64 pixels with 3 color channels for RGB).
MaxPooling2D((2, 2)):
Reduces the spatial dimensions by half, making the model computationally efficient and focusing on significant features.

2. Second Convolutional Block

Conv2D(64, (3, 3), activation='relu'):
Adds another convolutional layer with 64 filters.
MaxPooling2D((2, 2)):
Further reduces the spatial dimensions.

3. Third Convolutional Layer

Conv2D(128, (3, 3), activation='relu'):
Introduces a deeper layer with 128 filters to capture more complex features.

4. Flattening and Fully Connected Layers

Flatten():
Converts the 2D feature maps into a 1D vector to pass into fully connected layers.
Dense(128, activation='relu'):
Adds a dense layer with 128 neurons and ReLU activation for further learning.
Dropout(0.5):
Randomly drops 50% of neurons during training to prevent overfitting.

5. Output Layer

Dense(10, activation='softmax'):
Adds the final dense layer with 10 neurons (one for each class).
Uses the softmax activation function to output probabilities for each class.

6. Model Summary

Displays a summary of the model architecture, including the number of parameters for each layer and total parameters.

In [18]:
import tensorflow as tf
from tensorflow.keras import layers, models

# Creating the model
model = models.Sequential([
    # First convolutional layer and pooling layer
    layers.Conv2D(32, (3, 3), activation='relu', input_shape=(64, 64, 3)),  # 64x64 images and 3 color channels (RGB)
    layers.MaxPooling2D((2, 2)),
    # Second convolutional layer and pooling layer
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    # Third convolutional layer
    layers.Conv2D(128, (3, 3), activation='relu'),
    # Flatten for switching to fully connected layers
    layers.Flatten(),
    # Fully connected layer
    layers.Dense(128, activation='relu'),
    # Dropout layer
    layers.Dropout(0.5),
    # Output layer (softmax activation for 10 classes)
    layers.Dense(10, activation='softmax')
])

# View model summary
model.summary()

/usr/local/lib/python3.10/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                      │ (None, 62, 62, 32)          │             896 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d (MaxPooling2D)         │ (None, 31, 31, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_1 (Conv2D)                    │ (None, 29, 29, 64)          │          18,496 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_1 (MaxPooling2D)       │ (None, 14, 14, 64)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_2 (Conv2D)                    │ (None, 12, 12, 128)         │          73,856 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten (Flatten)                    │ (None, 18432)               │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense (Dense)                        │ (None, 128)                 │       2,359,424 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dropout (Dropout)                    │ (None, 128)                 │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 10)                  │           1,290 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 2,453,962 (9.36 MB)

 Trainable params: 2,453,962 (9.36 MB)

 Non-trainable params: 0 (0.00 B)

# 3. 2. Compiling the Model

optimizer='adam':
Uses the Adam optimizer, a widely used algorithm for fast and adaptive learning during backpropagation.
loss='categorical_crossentropy':
Suitable loss function for multi-class classification tasks where the target labels are one-hot encoded.
metrics=['accuracy']:
Tracks the model's accuracy as a performance metric during training and evaluation.

In [26]:
# compile the model
model.compile(
    optimizer='adam',  # Adam optimizer for fast learning
    loss='categorical_crossentropy',  # Suitable for multi-class classification
    metrics=['accuracy']  # Accuracy for performance evaluation
)

# 3.3 Training the Model

X_train_augmented:
Augmented training images are used to enhance model robustness and reduce overfitting.

tf.keras.utils.to_categorical(y_train_augmented, num_classes=10):
Converts integer labels to one-hot encoded format for compatibility with the loss function.
For example, label 3 becomes [0, 0, 0, 1, 0, 0, 0, 0, 0, 0].

validation_data:
Provides validation data (test images and one-hot encoded test labels) to monitor the model's performance on unseen data after each epoch.

epochs=20:
Trains the model for 20 iterations over the entire training dataset.

batch_size=32:
Divides the dataset into smaller batches of 32 samples for gradient updates, improving training efficiency.


In [27]:
# training the model
history = model.fit(
    X_train_augmented,  # Training images (augmented)
    tf.keras.utils.to_categorical(y_train_augmented, num_classes=10),  # Tags one-hot encoding
    validation_data=(
        X_test,
        tf.keras.utils.to_categorical(y_test, num_classes=10)
    ),  # Test data
    epochs=100,  # Epoch
    batch_size=16  # Batch size
)

Epoch 1/100
1707/1707 ━━━━━━━━━━━━━━━━━━━━ 17s 6ms/step - accuracy: 0.9508 - loss: 0.1601 - val_accuracy: 0.6856 - val_loss: 3.6281
Epoch 2/100
1707/1707 ━━━━━━━━━━━━━━━━━━━━ 13s 4ms/step - accuracy: 0.9464 - loss: 0.1725 - val_accuracy: 0.6718 - val_loss: 3.3247
Epoch 3/100
1707/1707 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9494 - loss: 0.1512 - val_accuracy: 0.6882 - val_loss: 3.3218
Epoch 4/100
1707/1707 ━━━━━━━━━━━━━━━━━━━━ 11s 4ms/step - accuracy: 0.9458 - loss: 0.1676 - val_accuracy: 0.6795 - val_loss: 3.1953
Epoch 5/100
1707/1707 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9477 - loss: 0.1749 - val_accuracy: 0.6677 - val_loss: 3.1404
Epoch 6/100
1707/1707 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9527 - loss: 0.1542 - val_accuracy: 0.6831 - val_loss: 2.9627
Epoch 7/100
1707/1707 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step - accuracy: 0.9461 - loss: 0.1723 - val_accuracy: 0.6600 - val_loss: 2.8993
Epoch 8/100
1707/1707 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9475 - los

for saving to drive and loading model

In [28]:
# create model save path
model_save_path = "/content/drive/My Drive/bootcamp_project/saved_model.h5"

# save the model
model.save(model_save_path)
print(f"Model saved at {model_save_path}")

from tensorflow.keras.models import load_model
# load the model
model = load_model("/content/drive/My Drive/bootcamp_project/saved_model.h5")


Model saved at /content/drive/My Drive/bootcamp_project/saved_model.h5


# 4. Testing the Model:

Tests the model on unseen data and reports accuracy and loss to measure generalization performance.

**model.evaluate**:
Evaluates the model's performance on the test dataset.
Returns the test loss and test accuracy.

**X_test**:
Test images are used for evaluation.

**tf.keras.utils.to_categorical(y_test, num_classes=10)**:
Converts test labels to one-hot encoded format.

**test_accuracy**:
Displays the model's accuracy on the test dataset as a percentage (e.g., 92.35%).

**test_loss**:
Outputs the loss value, representing the model's error on the test dataset.


In [29]:
# evaluating the model on the test dataset
test_loss, test_accuracy = model.evaluate(
    X_test,
    tf.keras.utils.to_categorical(y_test, num_classes=10)
)
print(f"Test Doğruluğu: {test_accuracy:.2%}")
print(f"Test Kaybı: {test_loss:.4f}")


61/61 ━━━━━━━━━━━━━━━━━━━━ 1s 7ms/step - accuracy: 0.6756 - loss: 5.5746
Test Doğruluğu: 68.00%
Test Kaybı: 5.2948


# 5. Manipulating Images with Different Lights:

In [43]:
# Color Constancy
"""
Global AI Image Processing Bootcamp
Dec 2024
Diclehan and Oguzhan Ulucan
"""
import numpy as np
import cv2

### Fundamental functions
def linearize_image(image):
    """
    Converts an sRGB image to linear RGB assuming the input image is in the range [0, 1]
    """
    return np.where(image <= 0.04045,
                    image / 12.92,
                    ((image + 0.055) / 1.055) ** 2.4)

def linear_to_srgb(image):
    """
    Converts a linear RGB image to sRGB assuming the input image is in the range [0, 1]
    """
    return np.where(image <= 0.0031308,
                    image * 12.92,
                    1.055 * (image ** (1 / 2.4)) - 0.055)

def handle_saturation(image, lower=0.05, upper=0.95):
    """
    Creates a mask for non-saturated pixels (those between `lower` and `upper` thresholds)
    """
    return np.all((image > lower) & (image < upper), axis=-1)

### Color constancy
def estimate_light_source_grey_world(image, mask):
    """
    Estimates the light source based on the Grey World assumption, using valid pixels from the mask
    """
    valid_pixels = image[mask]
    avg_color = np.mean(valid_pixels, axis=0)
    return avg_color / np.linalg.norm(avg_color)

def correct_colors(image, light_source):
    """
    Corrects the colors of the image by applying white balance using the estimated light source
    """
    return image * (1.0 / light_source)

def manipulate_light_source(image, light_color):
    """
    Simulates color manipulation under a different light source

    Args:
    - image: The input image (sRGB, [0, 1])
    - light_color: The light source color (unit norm RGB vector)

    Returns:
    - Manipulated image (sRGB, [0, 1])
    """
    # Step 1: Linearize the image
    linear_image = linearize_image(image)
    # Step 2: Apply the light source (multiplying the linear image by the light color)
    manipulated_image = linear_image * light_color
    # Step 3: Convert the manipulated image back to sRGB
    manipulated_srgb = linear_to_srgb(manipulated_image)

    return np.clip(manipulated_srgb, 0, 1)

def process_and_white_balance(image):
    """
    Applies white balance using both the Grey World and Max RGB methods
    Returns both corrected images in sRGB format
    """
    linear_image = linearize_image(image)
    valid_mask = handle_saturation(linear_image)

    # Light source estimations
    grey_world_light = estimate_light_source_grey_world(linear_image, valid_mask)
    # Color correction using both light sources
    corrected_grey_world = correct_colors(linear_image, grey_world_light)
    # Convert back to sRGB
    srgb_grey_world = linear_to_srgb(corrected_grey_world)
    # Clip and return
    return np.clip(srgb_grey_world, 0, 1)

# Light sources for color manipulation
def get_light_sources():
    """
    Returns a set of light sources for image manipulation
    """
    purplish_light = np.array([0.82, 0.15, 0.89]) / np.linalg.norm([0.82, 0.15, 0.89])
    yellowish_light = np.array([0.96, 0.24, 0.11]) / np.linalg.norm([0.96, 0.24, 0.11])
    greenish_light = np.array([0.11, 0.98, 0.12]) / np.linalg.norm([0.11, 0.98, 0.12])
    return purplish_light, yellowish_light, greenish_light


def get_wb_images(image):
    """
    Process and white balance the image
    """
    srgb_grey_world = process_and_white_balance(image)
    # Save the white-balanced images
    #cv2.imwrite('white_balanced_grey_world.jpg', cv2.cvtColor((srgb_grey_world*255).astype(np.uint8), cv2.COLOR_RGB2BGR))
    return srgb_grey_world

def get_manipulated_images(image):
    """
    Get manipulated images by applying color vectors
    """
    # Get the color vectors
    purplish_light, orangish_light, greenish_light = get_light_sources()

    # Manipulate the images under different light sources
    manipulated_purplish = manipulate_light_source(image, purplish_light)
    manipulated_orangish = manipulate_light_source(image, orangish_light)
    manipulated_greenish = manipulate_light_source(image, greenish_light)

    # Save the manipulated images
    #cv2.imwrite('manipulated_purplish.jpg', cv2.cvtColor((manipulated_purplish*255).astype(np.uint8), cv2.COLOR_RGB2BGR))
    #cv2.imwrite('manipulated_orangish.jpg', cv2.cvtColor((manipulated_orangish*255).astype(np.uint8), cv2.COLOR_RGB2BGR))
    #cv2.imwrite('manipulated_greenish.jpg', cv2.cvtColor((manipulated_greenish*255).astype(np.uint8), cv2.COLOR_RGB2BGR))
    return manipulated_purplish, manipulated_orangish, manipulated_greenish


In [48]:
# Create new test set for manipulation
X_test_manipulated = []
y_test_manipulated = []

for i, image in enumerate(X_test):
    # Give the normalized image to the manipulation function
    manipulated_purplish, manipulated_orangish, manipulated_greenish = get_manipulated_images(image)
    # Add manipulated images to list
    X_test_manipulated.append(manipulated_purplish)
    X_test_manipulated.append(manipulated_orangish)
    X_test_manipulated.append(manipulated_greenish)
    # Duplicate labels
    y_test_manipulated.extend([y_test[i]] * 3)

# convert from list to numpy array
X_test_manipulated = np.array(X_test_manipulated)
y_test_manipulated = np.array(y_test_manipulated)

print(f"Manipüle edilmiş X_test şekli: {X_test_manipulated.shape}")
print(f"Manipüle edilmiş y_test şekli: {y_test_manipulated.shape}")

Manipüle edilmiş X_test şekli: (5850, 64, 64, 3)
Manipüle edilmiş y_test şekli: (5850,)


# 6. Testing the Model with Manipulated Test Set:

In [50]:
# print manipulated dataset shape
print(f"X_test_manipulated şekli: {X_test_manipulated.shape}")
print(f"y_test_manipulated şekli: {y_test_manipulated.shape}")

# Model değerlendirme
y_test_manipulated_one_hot = tf.keras.utils.to_categorical(y_test_manipulated, num_classes=10)

# Modeli değerlendir
test_loss, test_accuracy = model.evaluate(X_test_manipulated, y_test_manipulated_one_hot)
print(f"Manipüle edilmiş test setindeki kayıp: {test_loss}")
print(f"Manipüle edilmiş test setindeki doğruluk: {test_accuracy}")

X_test_manipulated şekli: (5850, 64, 64, 3)
y_test_manipulated şekli: (5850,)
183/183 ━━━━━━━━━━━━━━━━━━━━ 1s 7ms/step - accuracy: 0.1862 - loss: 54.2937
Manipüle edilmiş test setindeki kayıp: 54.973899841308594
Manipüle edilmiş test setindeki doğruluk: 0.18905982375144958


# 7. Applying Color Constancy Algorithm to Manipulated Test Set:

In [51]:
def apply_white_balance_to_test_set(X_test_manipulated):
    """
    Apply white balance to all images in the manipulated test set.
    """
    wb_images = []
    for image in X_test_manipulated:
        # Apply white balance using get_wb_images
        wb_image = get_wb_images(image)
        wb_images.append(wb_image)
    return np.array(wb_images)

# apply
X_test_wb = apply_white_balance_to_test_set(X_test_manipulated)

# check the results
print(f"Renk sabitliği uygulanmış X_test_wb şekli: {X_test_wb.shape}")


/usr/local/lib/python3.10/dist-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
/usr/local/lib/python3.10/dist-packages/numpy/core/_methods.py:121: RuntimeWarning: invalid value encountered in divide
  ret = um.true_divide(


Renk sabitliği uygulanmış X_test_wb şekli: (5850, 64, 64, 3)


# 8. Testing the Model with the Color Consistency Applied Test Set:

In [52]:
# Evaluate the model on the color-consistent test set
test_loss_wb, test_accuracy_wb = model.evaluate(
    X_test_wb,
    tf.keras.utils.to_categorical(y_test_manipulated, num_classes=10)
)

print(f"Renk sabitliği uygulanmış test setindeki kayıp: {test_loss_wb}")
print(f"Renk sabitliği uygulanmış test setindeki doğruluk: {test_accuracy_wb}")


183/183 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.6468 - loss: 5.6075
Renk sabitliği uygulanmış test setindeki kayıp: 5.261206150054932
Renk sabitliği uygulanmış test setindeki doğruluk: 0.6529914736747742


# 9. Comparing and Reporting the Success of Different Test Sets:

**1. Initial Test Data Results**

Accuracy: 67.56%
Loss: 5.5746

The results on the initial test data show an accuracy of 67.56% and a loss of 5.5746. This indicates that while the model is able to make reasonably accurate predictions, it is not perfect. The model's general performance is moderate, and there is potential for improvement through fine-tuning and addressing the issues that arise from the data distribution and model architecture.

**2. Manipulated Test Data Results**

Accuracy: 18.62%
Loss: 54.2937

The results on the manipulated test data show a significant drop in performance, with accuracy at just 18.62% and a loss of 54.2937. This low accuracy suggests that the manipulations, which alter the natural structure of the data, made it difficult for the model to make accurate predictions. The manipulations likely changed the image characteristics in ways that the model wasn't trained to handle effectively.

**3. White-Balanced Test Data Results**

Accuracy: 65.30%
Loss: 5.2612

The results on the white-balanced test data show an accuracy of 65.30% and a loss of 5.2612. The performance here is closer to the initial test data, indicating that the white-balancing algorithm helped reduce the negative impact of the manipulations. While still not perfect, the results suggest that applying a color correction technique like white balancing can improve the model's ability to handle manipulated images and restore a more balanced performance.


**Analysis of the Results and Possible Issues**

1. Impact of Manipulations on Model Performance

Manipulations typically change the inherent properties of the data, making it harder for the model to predict accurately. Some possible reasons for the poor performance on the manipulated test data include:

Data Distribution Mismatch: The model trained on the original dataset might not generalize well to manipulated images due to significant changes in the lighting, colors, or other features. The manipulation may have altered the distribution of the data in a way the model was not trained to handle.
Model Underfitting or Overfitting: The model might not have been trained with sufficient variety in the data to learn the effects of manipulations. This can result in the model underperforming when it encounters manipulated data.

2. Limitations of the White-Balancing Algorithm

While the white-balancing algorithm improved performance on manipulated data, it may not be sufficient in all cases. The improvement was notable but still did not bring the performance back to the level of the initial test data. Some factors contributing to the limited success of white balancing include:

Inadequate Algorithm: The white-balancing method may not have been strong enough to counteract all types of manipulation, especially if the changes in color and lighting were extreme or non-linear.
Complex Lighting Conditions: The white-balancing technique might struggle to account for complex or dynamic lighting conditions, reducing its effectiveness in real-world scenarios.

**Solution Suggestions**

Training with More Manipulated Data:

To improve the model’s resilience to manipulations, it would be beneficial to train it with more manipulated data. By including images with a variety of manipulations (e.g., color changes, lighting shifts), the model can better generalize and become more robust to such variations.
Additionally, data augmentation techniques can be applied during training to artificially increase the diversity of the dataset, such as through rotation, scaling, or introducing synthetic lighting changes.

Enhancing the White-Balancing Algorithm:

The current white-balancing algorithm might need to be more sophisticated. Exploring more advanced white-balancing techniques, such as those based on machine learning or deep learning models, could provide better results under diverse conditions.
For example, algorithms like color constancy methods or learning-based color correction techniques could be explored to improve color consistency and lighting corrections in the manipulated images.
Improving the Model Architecture and Hyperparameters:

The model architecture itself might need to be refined. Experimenting with deeper architectures, like ResNet or VGG, or leveraging transfer learning from pre-trained models might lead to better results.
Also, tuning the learning rate and batch size, as well as adjusting the number of epochs and early stopping parameters, may lead to better model performance.
Utilizing More Diverse Manipulations:

In addition to color manipulations, other forms of image augmentations (such as brightness, contrast, or saturation variations) can be included in the training data to better simulate real-world variations.
This would allow the model to handle a broader range of manipulations during inference.

Monitoring Model Performance Regularly:

It’s important to monitor the model’s performance on a validation set to identify signs of overfitting or underfitting. Regular evaluations during training can help catch issues early and adjust the model’s learning process accordingly.


**Conclusion**

The manipulated and white-balanced test data results highlight challenges in handling manipulated images, but also suggest that techniques like white balancing can mitigate some of these issues. However, the overall performance still falls short of expectations. Implementing the suggested improvements, such as training with more diverse manipulated data, improving the white-balancing algorithm, and refining the model architecture, will likely lead to better handling of manipulated and complex real-world data.